modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RecoverFailureDocumentProcessor.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Processor that remediates a failure document by:
+ * - Copying the original index (and routing if present) from ctx.document.* into the document metadata
+ * - Extracting the original source from ctx.document.source
+ * - Removing the "error" and "document" fields from the root
+ * - Restoring the original source fields at the root of the document
+ */
+public final class RecoverFailureDocumentProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "recover_failure_document";
+
+    RecoverFailureDocumentProcessor(String tag, String description) {
+        super(tag, description);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public IngestDocument execute(IngestDocument document) throws Exception {
+        if (document.hasField("document") == false) {
+            throw new IllegalArgumentException("field [document] not present as part of path [document]");
```

Comment:
I wonder if we could reword these early validation exception messages. I worry that an end user might see this and think that there is some kind of field path that is misconfigured somewhere, when in reality the failure document is just not what the processor expected. Maybe something like `failure store document has unexpected structure, missing required [document] field` and so on